src/null-safety/understanding-null-safety/index.md

```diff
@@ -0,0 +1,1410 @@
+you leave a function. So there are some additional compile errors:
+
+### Invalid returns
+
+If a function has a non-nullable return type, then every path through the
+function must reach a `return` statement that returns a value. Before null
+safety, Dart was pretty lax about missing returns. For example:
+
+```dart
+String missingReturn() {
+  // No return.
+}
+```
+
+If you analyze this, you get a gentle *hint* that *maybe* you forgot a return,
+but if not, no big deal. That's because if execution reaches the end of a
+function body then Dart implicitly returns `null`. Since every type is nullable,
+*technically* this function is safe, even though it's probably not what you
+want.
+
+With sound non-nullable types, this program is flat out wrong and unsafe. Under
+null safety, you get a compile error if a function with a non-nullable return
+type doesn't reliably return a value. By "reliably", I mean that the language
+analyzes all of the control flow paths through the function. As long as they all
+return something, it is satisfied. The analysis is pretty smart, so even this
+function is OK:
+
+```dart
+String alwaysReturns(int n) {
+  if (n == 0) {
+    return 'zero';
+  } else if (n < 0) {
+    throw ArgumentError('Negative values not allowed.');
+  } else {
+    if (n > 1000) {
+      return 'big';
+    } else {
+      return n.toString();
+    }
+  }
+}
+```
+
+We'll dive more deeply into the new flow analysis in the next section.
+
+### Uninitialized variables
+
+When you declare a variable, if you don't give it an explicit initializer, Dart
+default initializes the variable with `null`. That's convenient, but obviously
+totally unsafe if the variable's type is non-nullable. So we have to tighten
+things up for non-nullable variables:
+
+*   **Top level variable and static field declarations must have an
+    initializer.** Since these are lazy and can be accessed and assigned from
+    anywhere in the program, it's impossible for the compiler to guarantee that
+    the variable will be assigned a value before it gets read. So the only safe
+    option is to require the declaration itself to have an initializing
+    expression that produces a value of the right type:
+
+    ```dart
+    int topLevel = 0;
+
+    class SomeClass {
+      static int staticField = 0;
+    }
+    ```
+
+*   **Instance fields must either have an initializer at the declaration, use an
+    initializing formal, or be initialized in the constructor's initialization
+    list.** That's a lot of jargon. Here are the examples:
+
+    ```dart
+    class SomeClass {
+      int atDeclaration = 0;
+      int initializingFormal;
+      int initializationList;
+
+      SomeClass(this.initializingFormal)
+          : initializationList = 0;
+    }
+    ```
+
+    In other words, as long as the field has a value by the time you reach the
+    constructor body, you're good.
+
+*   Local variables are the fun one. A non-nullable local variable *doesn't*
```

Comment:
[nit] "are a fun one" reads to me like "the semantics are kind of crazy here". This is almost the opposite of what you're trying to say, I assume. 

```suggestion
*   With local variables, we can be a lot smarter. A non-nullable local variable *doesn't*
```

Reply:
Changed to "most flexible case".